File: gunkim/2024_Weekly_I_Solved/24_02/16401_과자나눠주기.java
import java.util.*;
import java.io.*;

public class Main {


    private int[] cookie;

    public void solution() throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int m = Integer.parseInt(st.nextToken());
        int n = Integer.parseInt(st.nextToken());

        cookie = new int[n];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            cookie[i] = Integer.parseInt(st.nextToken());
        }

        Arrays.sort(cookie);


        int left = 1;
        int right = cookie[n - 1];
        int answer = 0;
        while (left <= right) {

            int mid = (left + right) / 2;

            int count = 0; // 몇 개의 과자를 나눠줄 수 있는가
            for (int i = 0; i < n; i++) {
                count += cookie[i] / mid;
            }

            if (count >= m) {
                answer = Math.max(answer, mid);
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        System.out.println(answer);

    }


    public static void main(String[] args) throws Exception {
        new Main().solution();
    }

}

File: LeeYooseok/BOJ_오늘의 문제/2024/2024-02/2024-02-25/16401 과자 나눠주기.py
M, N = map(int, input().split())
stick_list = list(map(int, input().split()))

left, right = 1, 1e9

answer = 0
while left <= right:
    mid = (left + right) // 2

    num = 0
    for stick in stick_list:
        num += (stick // mid)

    if num >= M:
        left = mid + 1
        answer = max(answer, int(mid))
    else:
        right = mid - 1

print(answer)
